sys/log/full/src/log_cbmem.c

```diff
@@ -22,91 +22,108 @@
-
-    cbmem = (struct cbmem *) log->l_arg;
-
-    rc = cbmem_append(cbmem, buf, len);
-    if (rc != 0) {
-        goto err;
-    }
-
-    return (0);
-err:
-    return (rc);
-}
-
-static int
-log_cbmem_append_body(struct log *log, const struct log_entry_hdr *hdr,
+log_cbmem_append_body(struct log *log, struct log_entry_hdr *hdr,
                       const void *body, int body_len)
 {
     struct cbmem *cbmem;
-
     struct cbmem_scat_gath sg = {
-        .entries = (struct cbmem_scat_gath_entry[2]) {
+        .entries = (struct cbmem_scat_gath_entry[3]) {
             {
                 .flat_buf = hdr,
-                .flat_len = sizeof *hdr,
+                .flat_len = LOG_BASE_ENTRY_HDR_SIZE,
+            },
+            {
+                .flat_buf = hdr->ue_imghash,
+                .flat_len = 0,
             },
             {
                 .flat_buf = body,
                 .flat_len = body_len,
             },
         },
-        .count = 2,
+        .count = 3,
     };
 
+    if (hdr->ue_flags & LOG_FLAGS_IMG_HASH) {
+        sg.entries[1].flat_len = LOG_IMG_HASHLEN;
+    }
+
     cbmem = (struct cbmem *) log->l_arg;
 
     return cbmem_append_scat_gath(cbmem, &sg);
 }
 
 static int
-log_cbmem_append_mbuf(struct log *log, const struct os_mbuf *om)
-{
-    struct cbmem *cbmem;
-    int rc;
-
-    cbmem = (struct cbmem *) log->l_arg;
-
-    rc = cbmem_append_mbuf(cbmem, om);
-    if (rc != 0) {
-        goto err;
-    }
+log_cbmem_append(struct log *log, void *buf, int len)
+{   
+    uint16_t hdr_len;
 
-    return (0);
+    hdr_len = log_hdr_len(buf);
 
-err:
-    return (rc);
+    return log_cbmem_append_body(log, buf, (uint8_t *)buf + hdr_len,
+                                 len - hdr_len);
 }
 
 static int
-log_cbmem_append_mbuf_body(struct log *log, const struct log_entry_hdr *hdr,
-                           const struct os_mbuf *om)
+log_cbmem_append_mbuf_body(struct log *log, struct log_entry_hdr *hdr,
+                           struct os_mbuf *om)
 {
     struct cbmem *cbmem;
-
     struct cbmem_scat_gath sg = {
-        .entries = (struct cbmem_scat_gath_entry[2]) {
+        .entries = (struct cbmem_scat_gath_entry[3]) {
             {
                 .flat_buf = hdr,
-                .flat_len = sizeof *hdr,
+                .flat_len = LOG_BASE_ENTRY_HDR_SIZE,
+            },
+            {
+                .flat_buf = hdr->ue_imghash,
+                .flat_len = 0,
             },
             {
                 .om = om,
             },
         },
-        .count = 2,
+        .count = 3,
     };
 
+
+    if (hdr->ue_flags & LOG_FLAGS_IMG_HASH) {
+        sg.entries[1].flat_len = LOG_IMG_HASHLEN;
+    }
+
     cbmem = (struct cbmem *) log->l_arg;
 
     return cbmem_append_scat_gath(cbmem, &sg);
 }
 
+static int
+log_cbmem_append_mbuf(struct log *log, struct os_mbuf *om)
+{
+    struct os_mbuf m;
+    uint16_t mlen;
+    uint16_t hdr_len;
+    struct log_entry_hdr hdr;
+
+    mlen = os_mbuf_len(om);
```

Comment:
This should use OS_MBUF_PKTLEN() rather than os_mbuf_len(). We want to check the length of the entire mbuf chain (to see if the subsequent pull up will succeed), not the length of the first buffer in the chain.

Reply:
Not applicable here.